client/homebrew/editor/metadataEditor/metadataEditor.jsx

```diff
@@ -234,15 +272,37 @@ const MetadataEditor = createClass({
 			<div className='field description'>
 				<label>description</label>
 				<textarea value={this.props.metadata.description} className='value'
-					onChange={(e)=>this.handleFieldChange('description', e)} />
+					onChange={(e)=>{
+						this.handleFieldChange('description', [
+							{
+								type : 'maxLength',
+								reg  : '^.{0,5}$',
+								msg  : 'Max description length of 5 characters.'
+							}
+						], e);
+					}} />
 			</div>
 			<div className='field thumbnail'>
 				<label>thumbnail</label>
 				<input type='text'
 					value={this.props.metadata.thumbnail}
 					placeholder='my.thumbnail.url'
 					className='value'
-					onChange={(e)=>this.handleFieldChange('thumbnail', e)} />
+					onChange={(e)=>{
+						this.handleFieldChange('thumbnail', [
+							{
+								type : 'maxLength',
+								reg  : '^.{0,5}$',
+								msg  : 'Max URL length of 5 characters.'
+							},
+							{
+								type : 'wStart',
+								reg  : '^W',
+								msg  : 'URL must start with W'
+							}
```

Comment:
Please keep in mind that my experience with form validation comes from Angular, so a lot of my opinions stem from its built-in reactive forms design. 

* [Reactive Forms](https://angular.io/guide/reactive-forms)
* [Validation Functions](https://angular.io/guide/form-validation#validator-functions)

**Strong suggestion:** Generally speaking, I think it makes a lot of sense to store all of the field validation rules in a single place. In Angular's reactive forms, that would be on the form control itself, but with React since there isn't exactly a paradigm or class types for that I would lean more towards a constant object that exists outside the scope of the React component.

The primary reason for this is if for some reason we needed to manually trigger the change event for a field, we would need to include that field's validation rules. This could lead to duplication of those rules or we could just forget to send them or send the wrong ones which could end up with weird scenarios. If the rules are all in one object, we don't need to pass them around and the `handleFieldChange` method can just always check for validation rules for each field.

**Suggestion:** Consider using a structure similar to Angular's form validation functions instead of a pre-defined object with regex as the validator. The reason for this is primarily that regex is more difficult to understand than standard JS. As an example, something like the `maxLength` regex above can be replaced by `value.length <= 5`.

**Suggestion:** Update the `errs` "typed" array to be an object with the string list of errors. Ex:
```js
errs: {
    description: ['Max description length of 5 characters'],
    thumbnail: [],
}
```
I describe it as a "typed" array because it loosely expects the array's objects to conform to a specific "type" in its current design.

<hr/>

Here's a full example of all of what I'm suggesting!

Validators:
```js
const validators = {
    thumbnail: [
        (value)=>{
            return value?.length > 5 ? 'Max URL length of 5 characters' : null
        },
        (value)=>{
            return (value ?? '')[0] !== 'W' ? 'URL must start with W' : null
        }
    ]
}
```

Then the `handleFieldChange` function could be something like this:

```js
	handleFieldChange : function(name, e){
        const inputRules = validators[name] ?? [];
        // Call each input rule with the input's value, filter out falsy ones
		const errors = inputRules.map(rule => rule(e.target.value)).filter(Boolean)

        // Set the error state to the previous error state plus the current field's errors or undefined
        this.setState((prevState)=>({
		    errs : {
                ...(prevState.errs ?? {})
                [name]: errors.length > 0 ? errors : undefined
        }))

        // If the state has any truthy errors, don't send the update
        if (Object.values(this.state.errs ?? {}).filter(Boolean).length === 0) {
            this.props.onChange({
			    ...this.props.metadata,
			    [name] : e.target.value
		    });
        }
    }
```

Reply:
So i've read through the above and it makes sense to me.  I agree that setting the rules outside of the inputs is best and had thought about nesting those validators inside the props properties themselves but I have no idea if that would have worked/made sense.  

Is there any value in moving the validators object to a file outside of editPage.js?  Currently the only place to change the metadata is on that page, but perhaps in the future that would change, and having the validators as it's own file so it can be imported wherever would help?   For example, being allowed to change the title or published status from the User Page.

Also, should I just copy/paste your code into the appropriate spots on this PR or close this PR in favor of your own PR?

Reply:
I put the code purely to be an example bc I felt my comment wasn't entirely clear on its own. I wrote them straight in the GitHub editor and didn't test them, so I definitely wouldn't recommend just pasting it in but they are definitely a good template off of which you could make changes.

Reply:
So I'm back at this, and feel close to just getting the basic stuff down (without the changes to UI/actually displaying errors), but bumping into a frustrating issue.  

First, I pretty much copied what you have above but I'm pretty sure there was a missing comma and closing `}` in the setState portion.  I'm pretty sure I fixed it correctly but mentioning it in case I didn't.   Second, I moved the validation rules into it's own file-- still in the Metadata Editor folder just for now.

[Here is the branch I just pushed to GH](https://github.com/Gazook89/homebrewery/tree/Field-Validation-System-2)

**Current Problem:**  Any text entered specifically into the Thumbnail URL field results in a console error about changing a "Uncontrolled Input" into a "Controlled Input".  I understand the difference, but I don't see how the "thumbnail url" field is any different from the "Title" field for which I do not get any error.  Neither do I get an error when typing into the Description field.   I feel like all three fields are setup exactly the same, yet I get different results....?

Reply:
I think I need to change the field `value` to `defaultValue={this.props.metadata.thumbnail}` and so on for each field.  This populates the fields with the appropriate saved values initially, but thereafter does not set the value to match the changing props.    Without this change, it's going in circles...the value is set to the prop value, and the prop is set by the value, and once we added a bit of logic to prevent writing to the prop suddenly the value couldn't be changed either.

*(right?)*